tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand.h

```diff
@@ -16,20 +16,24 @@ inline void rand_init(uint seed) {
 }
 
 template <bool APPROXIMATION_MODE>
-inline void rand() {
+inline void rand(float from, float to) {
 #pragma GCC unroll 0
     for (int d = 0; d < 8; d++) {
+        // Generate random float
         TTI_SFPMOV(0, 9, p_sfpu::LREG0, 8);
+        // Unset sign bit and Set exponent to 127 to ensure the float is within the range [1, 2).
         TTI_SFPSETSGN(0, p_sfpu::LREG0, p_sfpu::LREG0, 1);
         TTI_SFPSETEXP(127, p_sfpu::LREG0, p_sfpu::LREG0, 1);
         // TTI_SFPADDI(0xbfa7, p_sfpu::LREG0, 0);
         TTI_SFPSTORE(0, 4, 3, 0);
 
-        vFloat from = 2.1;
-        vFloat to = 5;
-
+        // -1 to ensure the float is within the range [0, 1).
         vFloat rand_floats = dst_reg[0];
-        rand_floats = (rand_floats - 1) * (to - from) + from;
+        // rand_floats -= vConst1;
+
+        vFloat v_from = from;
+        vFloat v_to = to;
+        rand_floats = (rand_floats - vConst1) * (v_to - v_from) + v_from;
```

Comment:
You could calculate v_to - v_from outside the loop and save it in a register and avoid the recalculation every time inside the loop.  May be you could put it into one of those programmable constants in the init phase, instead of passing them as arguments.

Reply:
It seems like that programmable constants can only be created in the face loop.

<img width="1219" alt="image" src="https://github.com/user-attachments/assets/b94dc4e4-65b4-44b7-bf53-5823527cb061">

Reply:
It needs to be in the initialization I think, see file ckernel_sfpu_recip.h

```
template <bool APPROXIMATION_MODE>
inline void _init_reciprocal_()
{
    vConstFloatPrgm0 = 1.442695f; // ln2_recip
    vConstFloatPrgm1 = 2.0f;
}

} // namespace sfpu
} // namesp
```

Reply:
If you want to avoid doing it in the initialization then I think you need to pass the float value's bit patterns as uint32_t and then read them outside the loop and calculate the range, see ckernel_sfpu_dropout.h 

```
// probability should be between 0 - INT_MAX (signed)
// scale should be binary representation of a float32
template <bool APPROXIMATION_MODE, int ITERATIONS>
inline void _calculate_dropout_(const int iterations, uint probability, uint scale)
{
    // SFPU microcode

    FWLOG1("calculate_dropout() -- probability:%x", probability);
    FWLOG1("calculate_dropout() -- scale:%x", scale);

    TT_SFPLOADI(p_sfpu::LREG1, 10, scale&0xFFFF);
    TT_SFPLOADI(p_sfpu::LREG1, 8, scale>>16);
    TT_SFPLOADI(p_sfpu::LREG2, 10, probability&0xFFFF);
    TT_SFPLOADI(p_sfpu::LREG2, 8, probability>>16);
    #pragma GCC unroll 0
    for (int d = 0; d < iterations; d++) {

        ////////////////////////
        ```
        
        Here I am doing some pre computation outside the loop.

Reply:
Given I am not using immediate, I cannot use TTI_SFPLOADI, I have to use TT_SFPLOADI which is slower since trisc core cannot just pass it to the tensix core, it has to create the instruction. The instructions can be directly passed from the trisc core to tensix core if they all use immediates

Reply:
Thank for your explanation, I've update the ckernel file accordingly.